File: crates/forcefield/src/simulation.rs
//! Port of d3-force `simulation.js`, minus the timer and event dispatch:
//! the caller owns the loop and calls [`Simulation::tick`].

use crate::bodies::{Bodies, Node};
use crate::forces::Force;
use crate::lcg::Random;

const INITIAL_RADIUS: f64 = 10.0;

/// `Math.PI * (3 - Math.sqrt(5))`
fn initial_angle() -> f64 {
    std::f64::consts::PI * (3.0 - 5f64.sqrt())
}

pub struct Simulation {
    bodies: Bodies,
    alpha: f64,
    alpha_min: f64,
    alpha_decay: f64,
    alpha_target: f64,
    /// Stored as d3 stores it: `1 - velocityDecay`.
    velocity_decay: f64,
    /// Insertion-ordered, like the JS `Map`; replacing a name keeps its slot.
    forces: Vec<(String, Box<dyn Force>)>,
    random: Random,
}

impl Simulation {
    /// `forceSimulation(nodes)`: index the nodes and initialise unset
    /// positions on the phyllotaxis spiral.
    pub fn new(nodes: &[Node]) -> Simulation {
        let alpha_min: f64 = 0.001;
        let mut sim = Simulation {
            bodies: Bodies::from_nodes(nodes),
            alpha: 1.0,
            alpha_min,
            alpha_decay: 1.0 - alpha_min.powf(1.0 / 300.0),
            alpha_target: 0.0,
            velocity_decay: 0.6,
            forces: Vec::new(),
            random: Random::lcg(),
        };
        sim.initialize_nodes();
        sim
    }

    /// `n` nodes, all unset (spiral initialisation).
    pub fn with_count(n: usize) -> Simulation {
        Simulation::new(&vec![Node::UNSET; n])
    }

    fn initialize_nodes(&mut self) {
        let b = &mut self.bodies;
        let angle = initial_angle();
        for i in 0..b.len() {
            let (fx, fy) = (b.fixed[2 * i], b.fixed[2 * i + 1]);
            if !fx.is_nan() {
                b.pos[2 * i] = fx;
            }
            if !fy.is_nan() {
                b.pos[2 * i + 1] = fy;
            }
            if b.pos[2 * i].is_nan() || b.pos[2 * i + 1].is_nan() {
                let radius = INITIAL_RADIUS * (0.5 + i as f64).sqrt();
                let a = i as f64 * angle;
                b.pos[2 * i] = radius * a.cos();
                b.pos[2 * i + 1] = radius * a.sin();
            }
            if b.vel[2 * i].is_nan() || b.vel[2 * i + 1].is_nan() {
                b.vel[2 * i] = 0.0;
                b.vel[2 * i + 1] = 0.0;
            }
        }
    }

    /// `simulation.tick(iterations)`.
    pub fn tick(&mut self, iterations: usize) {
        for _ in 0..iterations {
            self.alpha += (self.alpha_target - self.alpha) * self.alpha_decay;

            for (_, force) in self.forces.iter_mut() {
                force.apply(&mut self.bodies, self.alpha, &mut self.random);
            }

            let b = &mut self.bodies;
            let vd = self.velocity_decay;
            for i in 0..b.len() {
                for axis in 0..2 {
                    let k = 2 * i + axis;
                    if b.fixed[k].is_nan() {
                        b.vel[k] *= vd;
                        b.pos[k] += b.vel[k];
                    } else {
                        b.pos[k] = b.fixed[k];
                        b.vel[k] = 0.0;
                    }
                }
            }
        }
    }

    /// Tick until `alpha < alphaMin`, as the JS timer loop would. Returns
    /// the number of ticks run. With `alphaDecay` 0 this never stops, so a
    /// `max_ticks` cap is required.
    pub fn run(&mut self, max_ticks: usize) -> usize {
        let mut n = 0;
        while self.alpha >= self.alpha_min && n < max_ticks {
            self.tick(1);
            n += 1;
        }
        n
    }

    // --- nodes ---------------------------------------------------------

    pub fn bodies(&self) -> &Bodies {
        &self.bodies
    }

    pub fn bodies_mut(&mut self) -> &mut Bodies {
        &mut self.bodies
    }

    pub fn len(&self) -> usize {
        self.bodies.len()
    }

    pub fn is_empty(&self) -> bool {
        self.bodies.is_empty()
    }

    /// `simulation.nodes(nodes)`: replace the node set and re-initialise
    /// every force.
    pub fn set_nodes(&mut self, nodes: &[Node]) {
        self.bodies = Bodies::from_nodes(nodes);
        self.initialize_nodes();
        self.reinitialize_forces();
    }

    fn reinitialize_forces(&mut self) {
        for (_, force) in self.forces.iter_mut() {
            force.initialize(&self.bodies, &mut self.random);
        }
    }

    // --- parameters ----------------------------------------------------

    pub fn alpha(&self) -> f64 {
        self.alpha
    }
    pub fn set_alpha(&mut self, v: f64) -> &mut Self {
        self.alpha = v;
        self
    }
    pub fn alpha_min(&self) -> f64 {
        self.alpha_min
    }
    pub fn set_alpha_min(&mut self, v: f64) -> &mut Self {
        self.alpha_min = v;
        self
    }
    pub fn alpha_decay(&self) -> f64 {
        self.alpha_decay
    }
    pub fn set_alpha_decay(&mut self, v: f64) -> &mut Self {
        self.alpha_decay = v;
        self
    }
    pub fn alpha_target(&self) -> f64 {
        self.alpha_target
    }
    pub fn set_alpha_target(&mut self, v: f64) -> &mut Self {
        self.alpha_target = v;
        self
    }
    /// The user-facing value (d3 default 0.4).
    pub fn velocity_decay(&self) -> f64 {
        1.0 - self.velocity_decay
    }
    pub fn set_velocity_decay(&mut self, v: f64) -> &mut Self {
        self.velocity_decay = 1.0 - v;
        self
    }

    /// `simulation.randomSource(fn)`: swap the random stream and
    /// re-initialise forces.
    pub fn set_random_source(&mut self, random: Random) -> &mut Self {
        self.random = random;
        self.reinitialize_forces();
        self
    }

    // --- forces --------------------------------------------------------

    /// `simulation.force(name, force)`: install (or replace in place) a
    /// named force, initialising it against the current nodes.
    pub fn add_force(&mut self, name: impl Into<String>, mut force: Box<dyn Force>) -> &mut Self {
        let name = name.into();
        force.initialize(&self.bodies, &mut self.random);
        if let Some(slot) = self.forces.iter_mut().find(|(n, _)| *n == name) {
            slot.1 = force;
        } else {
            self.forces.push((name, force));
        }
        self
    }

    /// `simulation.force(name, null)`.
    pub fn remove_force(&mut self, name: &str) -> Option<Box<dyn Force>> {
        let i = self.forces.iter().position(|(n, _)| n == name)?;
        Some(self.forces.remove(i).1)
    }

    pub fn force(&self, name: &str) -> Option<&dyn Force> {
        self.forces
            .iter()
            .find(|(n, _)| n == name)
            .map(|(_, f)| f.as_ref())
    }

    pub fn force_mut(&mut self, name: &str) -> Option<&mut Box<dyn Force>> {
        self.forces
            .iter_mut()
            .find(|(n, _)| n == name)
            .map(|(_, f)| f)
    }

    /// Re-run `initialize` on one force after changing its accessors.
    pub fn reinitialize_force(&mut self, name: &str) {
        if let Some(i) = self.forces.iter().position(|(n, _)| n == name) {
            self.forces[i].1.initialize(&self.bodies, &mut self.random);
        }
    }

    pub fn force_names(&self) -> impl Iterator<Item = &str> {
        self.forces.iter().map(|(n, _)| n.as_str())
    }

    // --- queries -------------------------------------------------------

    /// `simulation.find(x, y, radius)`: linear scan for the closest node.
    pub fn find(&self, x: f64, y: f64, radius: Option<f64>) -> Option<usize> {
        let mut radius = match radius {
            None => f64::INFINITY,
            Some(r) => r * r,
        };
        let mut closest = None;
        for i in 0..self.bodies.len() {
            let dx = x - self.bodies.x(i);
            let dy = y - self.bodies.y(i);
            let d2 = dx * dx + dy * dy;
            if d2 < radius {
                closest = Some(i);
                radius = d2;
            }
        }
        closest
    }
}

File: js/forcefield-sim/src/simulation.js
// A d3-force-shaped facade over the forcefield wasm simulation.
//
// Consumers written against d3-force use: simulation.nodes() → objects with
// live x/y/vx/vy/fx/fy; force(name, force) including custom JS forces;
// alpha*/velocityDecay getters and setters; restart()/stop(); on('tick.ns').
// This module provides all of that on top of the wasm `Simulation`, whose
// positions live in wasm linear memory and are read through Float64Array
// views without copying.
//
// The wasm module is injected (`new Simulation(wasm, nodes)`) rather than
// imported, so the consumer chooses how the .wasm is loaded (bundler,
// web, or node target of wasm-pack).

import { dispatch } from './dispatch.js'

const AXIS = { x: 0, y: 1 }
const FIELDS = new Set(['x', 'y', 'vx', 'vy', 'fx', 'fy'])

function lcg() {
  let s = 1
  return () => (s = (1664525 * s + 1013904223) % 4294967296) / 4294967296
}

const isFn = (v) => typeof v === 'function'
const constant = (v) => () => v

/** Evaluate a number-or-function accessor once per item, like d3 does at initialize. */
function perItem(accessor, items, fallback) {
  if (accessor == null) accessor = fallback
  const f = isFn(accessor) ? accessor : constant(+accessor)
  return Float64Array.from(items, (item, i) => {
    const v = +f(item, i, items)
    return Number.isNaN(v) ? NaN : v
  })
}

export class Simulation {
  constructor(wasm, nodes = []) {
    this._wasm = wasm
    this._core = null
    this._nodes = []
    this._proxies = []
    this._forces = new Map() // name → descriptor (built-in) | function (custom)
    this._event = dispatch('tick', 'end')
    this._random = lcg()
    this._running = false
    this._frame = null
    this._views = null
    this.nodes(nodes)
  }

  // --- memory views ----------------------------------------------------

  /** Views onto wasm memory; re-fetched when the buffer was detached by growth. */
  views() {
    const v = this._views
    if (v && v.pos.buffer.byteLength !== 0) return v
    return (this._views = { pos: this._core.positions(), vel: this._core.velocities(), fix: this._core.fixed() })
  }

  _proxy(i, data) {
    const sim = this
    const target = { ...data, index: i }
    for (const k of FIELDS) delete target[k]
    return new Proxy(target, {
      get(t, key) {
        if (typeof key === 'string' && FIELDS.has(key)) {
          const views = sim.views()
          if (key === 'x' || key === 'y') return views.pos[2 * i + AXIS[key]]
          if (key === 'vx' || key === 'vy') return views.vel[2 * i + AXIS[key[1]]]
          const f = views.fix[2 * i + AXIS[key[1]]]
          return Number.isNaN(f) ? undefined : f
        }
        return t[key]
      },
      set(t, key, value) {
        if (typeof key === 'string' && FIELDS.has(key)) {
          const views = sim.views()
          if (key === 'x' || key === 'y') views.pos[2 * i + AXIS[key]] = +value
          else if (key === 'vx' || key === 'vy') views.vel[2 * i + AXIS[key[1]]] = +value
          else views.fix[2 * i + AXIS[key[1]]] = value == null ? NaN : +value
          return true
        }
        t[key] = value
        return true
      },
      deleteProperty(t, key) {
        if (key === 'fx' || key === 'fy') { sim.views().fix[2 * i + AXIS[key[1]]] = NaN; return true }
        if (typeof key === 'string' && FIELDS.has(key)) return false
        return delete t[key]
      },
      has(t, key) {
        if (key === 'fx' || key === 'fy') return !Number.isNaN(sim.views().fix[2 * i + AXIS[key[1]]])
        return FIELDS.has(key) || key in t
      },
    })
  }

  // --- nodes -----------------------------------------------------------

  /**
   * Get the node proxies, or set the node list. Input objects may carry
   * x, y, vx, vy, fx, fy (unset → d3 semantics) plus any data; the returned
   * proxies expose the six simulation fields live from wasm memory and
   * everything else as ordinary properties.
   */
  nodes(_) {
    if (_ === undefined) return this._proxies
    this._nodes = _
    const n = _.length
    const pos = new Float64Array(2 * n)
    for (let i = 0; i < n; i++) {
      pos[2 * i] = _[i].x == null ? NaN : +_[i].x
      pos[2 * i + 1] = _[i].y == null ? NaN : +_[i].y
    }
    if (this._core) this._core.setNodes(pos)
    else this._core = this._wasm.Simulation.fromPositions(pos)
    this._views = null
    for (let i = 0; i < n; i++) {
      const d = _[i]
      if (d.vx != null || d.vy != null) this._core.setVelocity(i, +d.vx || 0, +d.vy || 0)
      if (d.fx != null || d.fy != null) this._core.setFixed(i, d.fx == null ? NaN : +d.fx, d.fy == null ? NaN : +d.fy)
    }
    this._proxies = _.map((d, i) => this._proxy(i, d))
    for (const name of this._forces.keys()) this._install(name)
    return this
  }

  // --- parameters ------------------------------------------------------

  alpha(_) { return _ === undefined ? this._core.alpha() : (this._core.setAlpha(+_), this) }
  alphaMin(_) { return _ === undefined ? this._core.alphaMin() : (this._core.setAlphaMin(+_), this) }
  alphaDecay(_) { return _ === undefined ? this._core.alphaDecay() : (this._core.setAlphaDecay(+_), this) }
  alphaTarget(_) { return _ === undefined ? this._core.alphaTarget() : (this._core.setAlphaTarget(+_), this) }
  velocityDecay(_) { return _ === undefined ? this._core.velocityDecay() : (this._core.setVelocityDecay(+_), this) }
  randomSource(_) { return _ === undefined ? this._random : (this._random = _, this) }

  // --- forces ----------------------------------------------------------

  /** d3 semantics: `force(name)` gets, `force(name, null)` removes, `force(name, f)` installs in place. */
  force(name, _) {
    if (arguments.length < 2) return this._forces.get(name)
    if (_ == null) {
      this._forces.delete(name)
      this._core.removeForce(name)
    } else {
      this._forces.set(name, _)
      this._install(name)
    }
    return this
  }

  _install(name) {
    const f = this._forces.get(name)
    const nodes = this._proxies
    if (isFn(f)) {
      // Custom JS force: initialize(nodes, random) then apply(alpha) per tick,
      // mutating the live proxies (i.e. wasm memory) directly.
      if (f.initialize) f.initialize(nodes, this._random)
      // Rust passes fresh views for the duration of the call; the proxies
      // read them through views(), so the force never re-enters wasm.
      this._core.forceCustom(name, (alpha, pos, vel, fix) => {
        this._views = { pos, vel, fix }
        f(alpha)
      })
      this._views = null
      return
    }
    f._attach(this, name)
    const c = this._core
    switch (f.type) {
      case 'manyBody':
        c.forceManyBody(name, perItem(f._strength, nodes, -30), f._theta, f._distanceMin, f._distanceMax)
        break
      case 'link': {
        const links = f._resolve(nodes)
        const src = Uint32Array.from(links, (l) => l.source.index)
        const tgt = Uint32Array.from(links, (l) => l.target.index)
        const distance = perItem(f._distance, links, 30)
        const strength = f._strength == null ? new Float64Array(0) : perItem(f._strength, links, 0)
        c.forceLink(name, src, tgt, distance, strength, f._iterations)
        break
      }
      case 'collide':
        c.forceCollide(name, perItem(f._radius, nodes, 1), f._strength, f._iterations)
        break
      case 'x':
        c.forceX(name, perItem(f._x, nodes, 0), perItem(f._strength, nodes, 0.1))
        break
      case 'y':
        c.forceY(name, perItem(f._y, nodes, 0), perItem(f._strength, nodes, 0.1))
        break
      case 'center':
        c.forceCenter(name, f._x, f._y, f._strength)
        break
      case 'radial':
        c.forceRadial(name, perItem(f._radius, nodes, 0), f._x, f._y, perItem(f._strength, nodes, 0.1))
        break
      default:
        throw new Error(`unknown force type: ${f.type}`)
    }
    this._views = null
  }

  // --- stepping --------------------------------------------------------

  tick(iterations = 1) {
    this._core.tick(iterations)
    return this
  }

  /** Run to alphaMin synchronously (bounded), without events. Returns ticks run. */
  run(maxTicks = 10_000) {
    return this._core.run(maxTicks)
  }

  restart() {
    if (this._running) return this
    this._running = true
    const schedule = typeof requestAnimationFrame === 'function'
      ? (fn) => { this._frame = requestAnimationFrame(fn) }
      : (fn) => { this._frame = setTimeout(fn, 16) }
    const step = () => {
      if (!this._running) return
      this._core.tick(1)
      this._event.call('tick', this)
      if (this._core.alpha() < this._core.alphaMin()) {
        this._running = false
        this._frame = null
        this._event.call('end', this)
        return
      }
      schedule(step)
    }
    schedule(step)
    return this
  }

  stop() {
    this._running = false
    if (this._frame != null) {
      if (typeof cancelAnimationFrame === 'function') cancelAnimationFrame(this._frame)
      clearTimeout(this._frame)
      this._frame = null
    }
    return this
  }

  on(typename, _) {
    if (arguments.length < 2) return this._event.on(typename)
    this._event.on(typename, _)
    return this
  }

  find(x, y, radius) {
    const i = this._core.find(x, y, radius)
    return i < 0 ? undefined : this._proxies[i]
  }

  /** Release wasm memory. The simulation is unusable afterwards. */
  free() {
    this.stop()
    this._core?.free()
    this._core = null
  }
}

// --- force descriptors: the d3 factory API, evaluated at install time ---

class Descriptor {
  constructor(type) { this.type = type; this._sim = null; this._name = null }
  _attach(sim, name) { this._sim = sim; this._name = name }
  _set(key, value) {
    this[key] = value
    if (this._sim) this._sim._install(this._name)
    return this
  }
  _accessor(key, args) { return args.length ? this._set(key, args[0]) : this[key] }
}

export function forceManyBody() {
  const f = new Descriptor('manyBody')
  Object.assign(f, { _strength: -30, _theta: 0.9, _distanceMin: 1, _distanceMax: Infinity })
  f.strength = function (...a) { return this._accessor('_strength', a) }
  f.theta = function (...a) { return this._accessor('_theta', a) }
  f.distanceMin = function (...a) { return this._accessor('_distanceMin', a) }
  f.distanceMax = function (...a) { return this._accessor('_distanceMax', a) }
  return f
}

export function forceLink(links = []) {
  const f = new Descriptor('link')
  Object.assign(f, { _links: links, _id: (d) => d.index, _distance: 30, _strength: null, _iterations: 1 })
  f.links = function (...a) { return this._accessor('_links', a) }
  f.id = function (...a) { return this._accessor('_id', a) }
  f.distance = function (...a) { return this._accessor('_distance', a) }
  f.strength = function (...a) { return this._accessor('_strength', a) }
  f.iterations = function (...a) { return this._accessor('_iterations', a) }
  /** Resolve ids to node proxies in place, as d3's link.initialize does. */
  f._resolve = function (nodes) {
    const byId = new Map(nodes.map((d, i) => [this._id(d, i, nodes), d]))
    const find = (id) => {
      const n = byId.get(id)
      if (!n) throw new Error(`node not found: ${id}`)
      return n
    }
    this._links.forEach((l, i) => {
      l.index = i
      if (typeof l.source !== 'object') l.source = find(l.source)
      else if (l.source.index == null || nodes[l.source.index] !== l.source) l.source = find(this._id(l.source, l.source.index, nodes))
      if (typeof l.target !== 'object') l.target = find(l.target)
      else if (l.target.index == null || nodes[l.target.index] !== l.target) l.target = find(this._id(l.target, l.target.index, nodes))
    })
    return this._links
  }
  return f
}

export function forceCollide(radius) {
  const f = new Descriptor('collide')
  Object.assign(f, { _radius: radius == null ? 1 : radius, _strength: 1, _iterations: 1 })
  f.radius = function (...a) { return this._accessor('_radius', a) }
  f.strength = function (...a) { return this._accessor('_strength', a) }
  f.iterations = function (...a) { return this._accessor('_iterations', a) }
  return f
}

export function forceX(x) {
  const f = new Descriptor('x')
  Object.assign(f, { _x: x == null ? 0 : x, _strength: 0.1 })
  f.x = function (...a) { return this._accessor('_x', a) }
  f.strength = function (...a) { return this._accessor('_strength', a) }
  return f
}

export function forceY(y) {
  const f = new Descriptor('y')
  Object.assign(f, { _y: y == null ? 0 : y, _strength: 0.1 })
  f.y = function (...a) { return this._accessor('_y', a) }
  f.strength = function (...a) { return this._accessor('_strength', a) }
  return f
}

export function forceCenter(x = 0, y = 0) {
  const f = new Descriptor('center')
  Object.assign(f, { _x: x, _y: y, _strength: 1 })
  f.x = function (...a) { return this._accessor('_x', a) }
  f.y = function (...a) { return this._accessor('_y', a) }
  f.strength = function (...a) { return this._accessor('_strength', a) }
  return f
}

export function forceRadial(radius, x = 0, y = 0) {
  const f = new Descriptor('radial')
  Object.assign(f, { _radius: radius, _x: x, _y: y, _strength: 0.1 })
  f.radius = function (...a) { return this._accessor('_radius', a) }
  f.x = function (...a) { return this._accessor('_x', a) }
  f.y = function (...a) { return this._accessor('_y', a) }
  f.strength = function (...a) { return this._accessor('_strength', a) }
  return f
}
